File: service/src/test/java/no/nav/melosys/service/kontroll/feature/arbeidutland/kontroll/ArbeidUtlandKontrollTest.java
package no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll;

import java.util.List;

import no.nav.melosys.domain.mottatteopplysninger.MottatteOpplysningerData;
import no.nav.melosys.domain.mottatteopplysninger.data.ForetakUtland;
import no.nav.melosys.domain.mottatteopplysninger.data.arbeidssteder.FysiskArbeidssted;
import no.nav.melosys.service.validering.Kontrollfeil;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class ArbeidUtlandKontrollTest {

    private MottatteOpplysningerData mottatteOpplysningerData;

    @BeforeEach
    void setup() {
        mottatteOpplysningerData = new MottatteOpplysningerData();
        final var fysiskArbeidssted = new FysiskArbeidssted();
        fysiskArbeidssted.virksomhetNavn = " ";
        mottatteOpplysningerData.arbeidPaaLand.fysiskeArbeidssteder = List.of(fysiskArbeidssted);
        mottatteOpplysningerData.foretakUtland = List.of(new ForetakUtland());
    }

    @Test
    void utførKontroller_arbeidsstedManglerFelter_returnererKode() {
        Kontrollfeil kontrollfeil = ArbeidUtlandKontroll.arbeidsstedManglerFelter(mottatteOpplysningerData);
        assertThat(kontrollfeil)
            .extracting(Kontrollfeil::getKode, Kontrollfeil::getFelter)
            .contains(List.of(String.format(ArbeidUtlandKontroll.ARBEIDSSTED_FIRMANAVN, 0), String.format(ArbeidUtlandKontroll.ARBEIDSSTED_LAND, 0)));
    }

    @Test
    void utførKontroller_foretakUtlandManglerFelter_returnererKode() {
        Kontrollfeil kontrollfeil = ArbeidUtlandKontroll.foretakUtlandManglerFelter(mottatteOpplysningerData);
        assertThat(kontrollfeil)
            .extracting(Kontrollfeil::getKode, Kontrollfeil::getFelter)
            .contains(List.of(String.format(ArbeidUtlandKontroll.FORETAK_UTLAND_NAVN, 0), String.format(ArbeidUtlandKontroll.FORETAK_UTLAND_LAND, 0)));
    }
}

File: service/src/test/kotlin/no/nav/melosys/service/kontroll/ferdigbehandling/ArbeidUtlandKontrollTest.kt
package no.nav.melosys.service.kontroll.ferdigbehandling

import no.nav.melosys.domain.kodeverk.begrunnelser.Fartsomrader
import no.nav.melosys.domain.kodeverk.begrunnelser.Kontroll_begrunnelser
import no.nav.melosys.domain.mottatteopplysninger.MottatteOpplysningerData
import no.nav.melosys.domain.mottatteopplysninger.data.ForetakUtland
import no.nav.melosys.domain.mottatteopplysninger.data.arbeidssteder.FysiskArbeidssted
import no.nav.melosys.domain.mottatteopplysninger.data.arbeidssteder.LuftfartBase
import no.nav.melosys.domain.mottatteopplysninger.data.arbeidssteder.MaritimtArbeid
import no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll.ArbeidUtlandKontroll.Companion.arbeidsstedLandManglerFelter
import no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll.ArbeidUtlandKontroll.Companion.foretakUtlandManglerFelter
import no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll.ArbeidUtlandKontroll.Companion.luftfartArbeidsstedManglerFelter
import no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll.ArbeidUtlandKontroll.Companion.maritimtArbeidsstedManglerFelter
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test


internal class ArbeidUtlandKontrollTest {
    private var mottatteOpplysningerData: MottatteOpplysningerData? = null

    @BeforeEach
    fun setup() {
        mottatteOpplysningerData = MottatteOpplysningerData()
        val fysiskArbeidssted = FysiskArbeidssted()
        val maritimtArbeid = MaritimtArbeid()
        val luftfartBase = LuftfartBase()
        val foretakUtland = ForetakUtland();
        fysiskArbeidssted.virksomhetNavn = " "
        maritimtArbeid.enhetNavn = " "
        luftfartBase.hjemmebaseNavn = " "
        foretakUtland.selvstendigNæringsvirksomhet = false
        maritimtArbeid.fartsomradeKode = Fartsomrader.INNENRIKS
        mottatteOpplysningerData!!.arbeidPaaLand.fysiskeArbeidssteder = listOf(fysiskArbeidssted)
        mottatteOpplysningerData!!.maritimtArbeid = listOf(maritimtArbeid)
        mottatteOpplysningerData!!.luftfartBaser = listOf(luftfartBase)
        mottatteOpplysningerData!!.foretakUtland = listOf(foretakUtland)
    }

    @Test
    fun utførKontroller_arbeidsstedLandManglerFelter_returnererKode() {
        val kontrollfeil = arbeidsstedLandManglerFelter(mottatteOpplysningerData!!)!!
        assert(kontrollfeil.kode == Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSSTED_LAND)
    }

    @Test
    fun utførKontroller_arbeidsstedMaritimManglerFelter_returnererKode() {
        val kontrollfeil = maritimtArbeidsstedManglerFelter(mottatteOpplysningerData!!)!!
        assert(kontrollfeil.kode == Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSSTED_MARITIM)
    }

    @Test
    fun utførKontroller_arbeidsstedFlyManglerFelter_returnererKode() {
        val kontrollfeil = luftfartArbeidsstedManglerFelter(mottatteOpplysningerData!!)!!
        assert(kontrollfeil.kode == Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSSTED_LUFTFART)
    }

    @Test
    fun utførKontroller_foretakUtlandManglerFelter_returnererKode() {
        val kontrollfeil = foretakUtlandManglerFelter(mottatteOpplysningerData!!)!!
        assert(kontrollfeil.kode == Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSFORHOLD_UTL)
    }
}
